### PR TITLE
Implement support for returning metadata about a domain

### PR DESCRIPTION
**Pull request by hellais**
The category code is the one of the shortest URL in the test list (with
priority to the URL in the global list)

The canonical URL is that of whatever is inside of the global or if not
present there, then it's the one of the shortest variation of the domain

### CODE REVIEW

**Comment by hellais** [comment]
This fixes: https://github.com/ooni/backend/issues/609

File: newapi/ooniapi/private.py [hunk 1]
```diff
@@ -843,3 +841,55 @@ def api_private_circumvention_runtime_stats() -> Response:
 
     except Exception as e:
         return jsonify({"v": 0, "error": str(e)})
+
+
+@api_private_blueprint.route("/domain_metadata")
+def api_private_domain_metadata() -> Response:
+    domain = request.args.get("domain")
```
Comment (bassosimone) [inline newapi/ooniapi/private.py:848]: Please, include an explanatory message to this algorithm where:

1. you explain the problem we're trying to solve in terms of (a) canonical category of a domain; (b) managing www (the ideal here would link to the issue we're solving here, which I don't know which is)

2. you refer to the your gist where you investigated the test list

File: newapi/ooniapi/private.py [hunk 2]
```diff
@@ -843,3 +841,55 @@ def api_private_circumvention_runtime_stats() -> Response:
...
+        domain_variations.append(domain.lstrip("www."))
+    else:
+        domain_variations.append("www." + domain)
+
+    category_code = "MISC"
+    canonical_domain = sorted(domain_variations, key=len)[0]
+
+    params = {str(idx): d for idx, d in enumerate(domain_variations)}
+    domain_q = "OR ".join([f"domain = %({idx})s " for idx in params.keys()])
+    res = query_click_one_row(
```
Comment (bassosimone) [inline newapi/ooniapi/private.py:863]:
```suggestion
    # case 1: domain with or without www is in the global list (cc = 'ZZ')
    res = query_click_one_row(
```

File: newapi/ooniapi/private.py [hunk 3]
```diff
@@ -843,3 +841,55 @@ def api_private_circumvention_runtime_stats() -> Response:
...
+        FROM citizenlab 
+        WHERE ({domain_q})
+        AND cc = 'ZZ' 
+        ORDER BY length(url)
+        LIMIT 1
+    """,
+        params,
+    )
+    if not res:
+        res2 = query_click_one_row(
```
Comment (bassosimone) [inline newapi/ooniapi/private.py:875]:
```suggestion
        # case 2: domain only inside a country list, so we just select the
        # shortest domain among the one with and without www
        res2 = query_click_one_row(
```

**Review by bassosimone — APPROVED** [review]
We discussed this change extensively. The proposed solution addresses the two original problems with domain-centric pages (which however still need to be documented in the code before merging, so we know why we added this function when doing future code review and we have a reference to the whole problem).

I do not see any performance issue or impact with respect to the Clickhouse backend given that we're querying a table that is very small by Clickhouse standards.

I do see however issues in the test lists becoming inconsistent unless we apply a CI check that ensures someone does not add https://foobar.com to a country list when https://foobar.com/antani is already in the test list. I think we should at the very minimum create an issue about this topic before merging.

To summarize, please before merging make sure that:

1. we have better in-code documentation
2. we have an issue (or a fix) regarding CI checks in the test lists

Thank you for moving this forward!

🐳

File: newapi/ooniapi/private.py (hunk 1 above)
Comment (hellais) [reply newapi/ooniapi/private.py:871]: Added more doc strings and a link to the gist

**Comment by hellais** [comment]
Thanks @FedericoCeratto and @bassosimone for the review and input.

I documented what needs to be done to address the issues at the citizenlab test-lists level: https://github.com/ooni/backend/issues/611. I don't think we need to do have this be done prior to merge and deploy. The consequence is just that some categories will be for some amount of less precise.
 
For me we can proceed with merging and deploying this.
